### ISSUE TITLE
Bug: Incorrect loss initialization and assignment in ModularTorchModel and GroverModel

### ISSUE THREAD

**Issue by anushkagupta200615-jpg** [deepchem/deepchem#5045]
In `deepchem/models/torch_models/modular.py` (around line 84), the `ModularTorchModel` initializes the super class (`TorchModel`) by passing `self.loss_func` to the `loss` parameter:

```python
# FIXME self.loss_func is an incorrect argument for TorchModel.loss because
# it performs more than computing loss
super().__init__(self.model, self.loss_func, **kwargs)
```

However, later in the `fit` method of `ModularTorchModel` (around line 193), this causes redundancy and confusion:

```python
# FIXME This line is not needed as loss is computed inside the call to loss_func
if loss is None:
    loss = self._loss_fn
```

Furthermore, in `deepchem/models/torch_models/grover.py` (around line 412), `GroverModel` blindly overwrites this attribute immediately after calling `super().__init__()`:

```python
# FIXME In the above step, we initialize modular torch model but
# something is missing here. The attribute loss from TorchModel gets assigned `loss_func`
# by super class initialization in ModularTorchModel but here we reinitialize it.
self.loss = self.get_loss_func()
```

## To Reproduce

1. Navigate to `deepchem/models/torch_models/modular.py` at line 84 and line 193.
2. Navigate to `deepchem/models/torch_models/grover.py` at line 412.
3. Observe the conflicting class initialization logic and overwriting of the `self.loss` attribute.

## Expected behavior

`ModularTorchModel` and `GroverModel` should properly distinguish between `loss` (a tensor or value) and `loss_func` (the callable method that calculates it). 

The `fit()` loop in `ModularTorchModel` needs to be refactored to rely on the correct loss computation pipeline. The `GroverModel` should be updated to cleanly hook into the parent class's initialization without needing to haphazardly overwrite the `self.loss` attribute post-initialization.

## Environment

* OS: Any
* Python version: Any
* DeepChem version: Master branch (latest)
* RDKit version (optional): N/A
* TensorFlow version (optional): N/A
* PyTorch version (optional): Any
* Any other relevant information: This is a static code architecture issue within the PyTorch models directory.

## Additional context

Fixing this structural issue will prevent silent failures or unexpected loss computation behaviors when users subclass `ModularTorchModel` or try to provide custom loss overrides.
